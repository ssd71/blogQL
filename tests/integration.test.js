const { createTestClient } = require('apollo-server-testing');
const gql = require('graphql-tag');
// const { graphqlserver } = require('../');
const { constructTestServer } = require('./__utils__');


const mockContext = {
  PostAPI: {
    getAllPosts: jest.fn(),
    getPost: jest.fn(),
    createPost: jest.fn(),
    updatePost: jest.fn(),
    deletePost: jest.fn(),
  },
};

const testPost = {
  id: 1,
  title: 'testTitle',
  description: 'testDesc',
  body: 'testBody',
};

const GET_ALL_POSTS = gql`
  query {
    feed {
      id
      description
      title
      body
    }
  }
`;

const GET_POST = gql`
  query getPost($id: Int!) {
    getPost(id: $id){
      id
      title
      description
      body
    }
  }
`;

const UPDATE_POST = gql`
  mutation updatePost($id: Int!) {
    updatePost(id: $id, title:"this sucks")
  }
`;

const CREATE_POST = gql`
mutation createPost($title: String!, $description: String, $body: String!) {
    createPost(title: $title, description: $description, body: $body) {
      id
    }
  }
`;

const DELETE_POST = gql`
  mutation deletePost($id: Int!) {
    deletePost(id: $id)
  }
`;

describe('Queries', () => {
  it('gets all posts', async () => {
    mockContext.PostAPI.getAllPosts.mockReturnValueOnce([testPost]);
    const server = constructTestServer(mockContext);
    const { query } = createTestClient(server);
    const res = await query({ query: GET_ALL_POSTS });
    expect(res).toMatchSnapshot();
  });

  it('gets a single post', async () => {
    mockContext.PostAPI.getPost.mockReturnValueOnce(testPost);
    const server = constructTestServer(mockContext);
    const { query } = createTestClient(server);
    const res = await query({ query: GET_POST, variables: { id: 1 } });
    expect(res).toMatchSnapshot();
  });
});

describe('Mutations', () => {
  it('creates single post', async () => {
    mockContext.PostAPI.createPost.mockReturnValueOnce(testPost);
    const server = constructTestServer(mockContext);
    const { query } = createTestClient(server);
    const res = await query({
      query: CREATE_POST,
      variables: {
        title: testPost.title, body: testPost.body,
      },
    });
    expect(res).toMatchSnapshot();
  });

  it('updates existing post', async () => {
    mockContext.PostAPI.updatePost.mockImplementation(id => (id === 1 ? [1] : [0]));
    const server = constructTestServer(mockContext);
    const { query } = createTestClient(server);
    const res = await query({ query: UPDATE_POST, variables: { id: 1 } });
    expect(res).toMatchSnapshot();
  });

  it('deletes single post', async () => {
    mockContext.PostAPI.deletePost.mockImplementation(id => (id === 1 ? 1 : 0));
    const server = constructTestServer(mockContext);
    const { query } = createTestClient(server);
    const res = await query({ query: DELETE_POST, variables: { id: 1 } });
    expect(res).toMatchSnapshot();
  });
});
