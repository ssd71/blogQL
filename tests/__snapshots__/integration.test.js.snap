// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutations creates single post 1`] = `
Object {
  "data": Object {
    "createPost": Object {
      "id": 1,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutations deletes single post 1`] = `
Object {
  "data": Object {
    "deletePost": 0,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Mutations updates existing post 1`] = `
Object {
  "data": Object {
    "updatePost": Array [
      0,
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries gets a single post 1`] = `
Object {
  "data": Object {
    "getPost": Object {
      "body": "testBody",
      "description": "testDesc",
      "id": 1,
      "title": "testTitle",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries gets all posts 1`] = `
Object {
  "data": Object {
    "feed": Array [
      Object {
        "body": "testBody",
        "description": "testDesc",
        "id": 1,
        "title": "testTitle",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
